src/Controller/InAppPurchasesController.php

```diff
@@ -0,0 +1,138 @@
+use Shopware\Core\Framework\Validation\DataBag\RequestDataBag;
+use SwagExtensionStore\Exception\ExtensionStoreException;
+use SwagExtensionStore\Services\InAppPurchasesService;
+use SwagExtensionStore\Struct\InAppPurchaseCartPositionCollection;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Attribute\Route;
+
+/**
+ * @internal
+ */
+#[Package('checkout')]
+#[Route(defaults: ['_routeScope' => ['api']])]
+class InAppPurchasesController
+{
+    /**
+     * @param EntityRepository<AppCollection> $appRepository
+     */
+    public function __construct(
+        private readonly InAppPurchasesService $inAppPurchasesService,
+        private readonly InAppPurchaseUpdater $inAppPurchaseUpdater,
+        private readonly AbstractExtensionDataProvider $extensionDataProvider,
+        private readonly InAppPurchasesGateway $appPurchasesGateway,
+        private readonly EntityRepository $appRepository,
+    ) {}
+
+    #[Route('/api/_action/in-app-purchases/{technicalName}/details', name: 'api.in-app-purchases.detail', methods: ['GET'])]
+    public function getInAppFeature(string $technicalName, Context $context): Response
+    {
+        $criteria = new Criteria();
+        $criteria->addFilter(new EqualsFilter('name', $technicalName));
+
+        $extension = $this->extensionDataProvider
+            ->getInstalledExtensions($context, false, $criteria)
+            ->first();
+
+        return new JsonResponse($extension);
+    }
+
+    #[Route('/api/_action/in-app-purchases/cart/new', name: 'api.in-app-purchases.cart.new', methods: ['POST'])]
+    public function createCart(RequestDataBag $data, Context $context): Response
+    {
+        $name = $data->getString('name');
+        $feature = $data->getString('feature');
+
+        $cart = $this->inAppPurchasesService->createCart($name, $feature, $context);
+
+        return new JsonResponse($cart);
+    }
+
+    #[Route('/api/_action/in-app-purchases/cart/order', name: 'api.in-app-purchases.cart.order', methods: ['POST'])]
+    public function orderCart(RequestDataBag $data, Context $context): Response
+    {
+        $taxRate = \floatval($data->getString('taxRate'));
+        $positions = $data->get('positions');
+        $extensionName = $data->get('name');
+
+        $positionCollection = InAppPurchaseCartPositionCollection::fromArray($positions->all());
+
+        $app = $this->getAppByName($extensionName, $context);
+        if (!$app) {
+            return $this->inAppPurchasesService->orderCart($taxRate, $positionCollection->toCart(), $context);
+        }
+
+        $payload = new InAppPurchasesPayload($positionCollection->getIdentifiers());
+        $validCartItems = $this->appPurchasesGateway->process($payload, $context, $app);
+        if (!$validCartItems) {
+            return $this->inAppPurchasesService->orderCart($taxRate, $positionCollection->toCart(), $context);
```

Comment:
why do we order when there are no cart items? can't we throw?

```
throw ExtensionStoreException::invalidInAppPurchase();
```

Reply:
Since an app gateway url is not a requirement, it would return null if no app gateway url was found. therefor we proceed with the order.

I could check app gateway URL here as well, and then throw an exception if something else goes wrong?

i.e:

```
if ($appGatewayUrl && !$validCartItems) {
throw exception;
}
```

Reply:
ah, it wasn't clear to me that `$validCartItems === null` means that there was no gateway to be called.
Maybe just add a comment like
```
// no response means no filtering by app
```
and rename the variable to `$iapGatewayResponse`. I assumed, without looking at the `InAppPurchasesGateway` class that this was an array of line items, and therefore `!$validCartItems` meant, that they were all filtered by the app.